framework/src/bundle/CoreBundleContext.cpp

```diff
@@ -169,20 +169,32 @@ void CoreBundleContext::Uninit1()
   listeners.Clear();
   resolver.Clear();
 
+  // Do not hold the bundleThreads lock while calling
+  // BundleThread::Quit() because this will join the bundle
+  // thread which may need to acquire the bundleThreads lock
+  // itself.
+  //
+  // At this point, all bundles are stopped and all
+  // bundle context instances invalidated. So no new bundle
+  // threads can be created and it is sufficient to get
+  // the current list once and not check for new bundle
+  // threads again.
+  std::list<std::shared_ptr<BundleThread>> threads;
+  bundleThreads.Lock(), std::swap(threads, bundleThreads.value);
```

Comment:
What I am saying is that at this point, the only threads left who may modify `bundleThreads.value` (or `zombies`) are the ones calling `Framework::WaitForStop` and they are properly synchronized. Non-zombie bundle threads which have not been joined yet are waiting for an operation (an hence are elements of the `bundleThreads.value` list) and we are joining them explicitly. No new bundle threads can be created and no operation can be queued at this point.